dvc/tree/dropbox.py

```diff
@@ -0,0 +1,466 @@
+            EXPIRATION: os.environ.get(EXPIRATION, ""),
+            REFRESH_TOKEN: os.environ[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        _creds = creds.copy()
+        _creds[EXPIRATION] = _creds[EXPIRATION].isoformat()
+        os.environ.update(_creds)
+
+
+class FileCredProvider(DropboxCredProvider):
+    DEFAULT_FILE = "dropbox-user-credentials.json"
+    DROPBOX_CREDENTIALS_FILE = "DROPBOX_CREDENTIALS_FILE"
+
+    def __init__(self, repo):
+        self._cred_location = os.environ.get(
+            self.DROPBOX_CREDENTIALS_FILE,
+            os.path.join(repo.tmp_dir, self.DEFAULT_FILE),
+        )
+
+    def can_auth(self):
+        if not os.path.exists(self._cred_location):
+            return False
+        with open(self._cred_location) as infile:
+            return REFRESH_TOKEN in json.load(infile)
+
+    def _gather_auth(self):
+        with open(self._cred_location) as infile:
+            config = json.load(infile)
+        return {
+            ACCESS_TOKEN: config.get(ACCESS_TOKEN, ""),
+            EXPIRATION: config.get(EXPIRATION, ""),
+            REFRESH_TOKEN: config[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        _creds = creds.copy()
+        _creds[EXPIRATION] = _creds[EXPIRATION].isoformat()
+        with open(self._cred_location, "w") as outfile:
+            json.dump(_creds, outfile)
+
+
+def path_info_to_dropbox_path(path_info):
+    return "/" + path_info.bucket + "/" + path_info.path
+
+
+class DropboxTree(BaseTree):
+    scheme = Schemes.DROPBOX
+    PATH_CLS = CloudURLInfo
+    REQUIRES = {
+        "dropbox": "dropbox",
+    }
+    PARAM_CHECKSUM = "content_hash"
+    # Per Dropbox API docs:
+    # This is an approximate number and there can be slightly more entries
+    # returned in some cases.
+    LIST_OBJECT_PAGE_SIZE = 1000
+    # Always prefer traverse for Dropbox since API usage quotas are a concern.
+    TRAVERSE_WEIGHT_MULTIPLIER = 1
+    TRAVERSE_PREFIX_LEN = 2
+
+    def __init__(self, repo, config):
+        super().__init__(repo, config)
+
+        self.path_info = self.PATH_CLS(config.get("url", "dropbox://default"))
+        self.chunk_size_mb = config.get("chunk_size", 48)
+
+        if self.chunk_size_mb % 4 != 0 or not 4 <= self.chunk_size_mb <= 150:
+            raise DvcException(
+                "Invalid chunk size for Dropbox client.\n"
+                "Chunk size must be between 4 and 150 MB and a multiple of 4."
+                " Was: {}".format(self.chunk_size_mb)
+            )
+
+    @wrap_prop(threading.Lock())
+    @cached_property
+    def client(self):
+        cred_prv = EnvCredProvider()
+        if cred_prv.can_auth():
+            logger.debug("Logging into Dropbox with environment variables")
+            return cred_prv.auth().client
+        cred_prv = FileCredProvider(self.repo)
+        if cred_prv.can_auth():
+            logger.debug("Logging into Dropbox with credentials file")
+            return cred_prv.auth().client
```

Comment:
I didn't spot the refresh token to change anytime. Therefore, actually you only need the refresh token to get the access token each time before the API call. The remaining elements could probably be dropped, but I introduced them to keep compatibility if they start to timeout the refresh token.

Reply:
yep, but it means we are going to keep refreshing them every run? (extra API call, extra time to an operation). Probably not the biggest problem, but I thought it can be a simple fix since you already have `save()` mechanics.

Reply:
The access token lives 4 hours. It's not much. We create the client once per dvc command being ran, so the update takes place at most once per CLI run.